src/flat.mjs

```diff
@@ -0,0 +1,39 @@
+import ensureIterable from './internal/ensure-iterable'
+
+const defaultShouldIFlat = (depth) => {
+  if (typeof depth === 'function') {
+    return depth
+  }
+  if (typeof depth === 'number') {
+    return (currentDepth, iter) =>
+      currentDepth <= depth && typeof iter[Symbol.iterator] === 'function' && typeof iter !== 'string'
+  }
+  throw new Error('flat: "depth" can be a function or a number')
+}
+
+function flat (shouldIFlat, iterable) {
+  function * _flat (currentDepth, iterable) {
+    if (shouldIFlat(currentDepth, iterable)) {
```

Comment:
I think you should remove iterable check [here](https://github.com/sithmel/iter-tools/blob/0b5d86e12d1950bac9599bd59a9c83992481e834/src/flat.mjs#L9) and move to this line, right before calling `shouldIFlat()` (`typeof iter[Symbol.iterator] === 'function' && shouldIFlat()`).

The current implementation burdens user to check if `iter` parameter is actually iterable (if the user doesn't, `flat()` will attempt to iterate on non-iterables causing errors).